File: 50_Horserace/python/horserace.py
import math
import random
import time


def basic_print(*zones, **kwargs) -> None:
    """Simulates the PRINT command from BASIC to some degree.
    Supports `printing zones` if given multiple arguments."""

    line = ""
    if len(zones) == 1:
        line = str(zones[0])
    else:
        line = "".join([f"{str(zone):<14}" for zone in zones])
    identation = kwargs.get("indent", 0)
    end = kwargs.get("end", "\n")
    print(" " * identation + line, end=end)


def basic_input(prompt, type_conversion=None):
    """BASIC INPUT command with optional type conversion"""

    while True:
        try:
            inp = input(f"{prompt}? ")
            if type_conversion is not None:
                inp = type_conversion(inp)
            break
        except ValueError:
            basic_print("INVALID INPUT!")
    return inp


# horse names do not change over the program, therefore making it a global.
# throught the game, the ordering of the horses is used to indentify them
HORSE_NAMES = [
    "JOE MAW",
    "L.B.J.",
    "MR.WASHBURN",
    "MISS KAREN",
    "JOLLY",
    "HORSE",
    "JELLY DO NOT",
    "MIDNIGHT",
]


def introduction():
    """Print the introduction, and optional the instructions"""

    basic_print("HORSERACE", indent=31)
    basic_print("CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY", indent=15)
    basic_print("\n\n")
    basic_print("WELCOME TO SOUTH PORTLAND HIGH RACETRACK")
    basic_print("                      ...OWNED BY LAURIE CHEVALIER")
    y_n = basic_input("DO YOU WANT DIRECTIONS")

    # if no instructions needed, return
    if y_n.upper() == "NO":
        return

    basic_print("UP TO 10 MAY PLAY.  A TABLE OF ODDS WILL BE PRINTED.  YOU")
    basic_print("MAY BET ANY + AMOUNT UNDER 100000 ON ONE HORSE.")
    basic_print("DURING THE RACE, A HORSE WILL BE SHOWN BY ITS")
    basic_print("NUMBER.  THE HORSES RACE DOWN THE PAPER!")
    basic_print("")


def setup_players():
    """Gather the number of players and their names"""

    # ensure we get an integer value from the user
    number_of_players = basic_input("HOW MANY WANT TO BET", int)

    # for each user query their name and return the list of names
    player_names = []
    basic_print("WHEN ? APPEARS,TYPE NAME")
    for _ in range(number_of_players):
        player_names.append(basic_input(""))
    return player_names


def setup_horses():
    """Generates random odds for each horse. Returns a list of
    odds, indexed by the order of the global HORSE_NAMES."""

    odds = [random.randrange(1, 10) for _ in HORSE_NAMES]
    total = sum(odds)

    # rounding odds to two decimals for nicer output,
    # this is not in the origin implementation
    return [round(total / odd, 2) for odd in odds]


def print_horse_odds(odds) -> None:
    """Print the odds for each horse"""

    basic_print("")
    for i in range(len(HORSE_NAMES)):
        basic_print(HORSE_NAMES[i], i, f"{odds[i]}:1")
    basic_print("")


def get_bets(player_names):
    """For each player, get the number of the horse to bet on,
    as well as the amount of money to bet"""

    basic_print("--------------------------------------------------")
    basic_print("PLACE YOUR BETS...HORSE # THEN AMOUNT")

    bets = []
    for name in player_names:
        horse = basic_input(name, int)
        amount = None
        while amount is None:
            amount = basic_input("", float)
            if amount < 1 or amount >= 100000:
                basic_print("  YOU CAN'T DO THAT!")
                amount = None
        bets.append((horse, amount))

    basic_print("")

    return bets


def get_distance(odd):
    """Advances a horse during one step of the racing simulation.
    The amount travelled is random, but scaled by the odds of the horse"""

    d = random.randrange(1, 100)
    s = math.ceil(odd)
    if d < 10:
        return 1
    elif d < s + 17:
        return 2
    elif d < s + 37:
        return 3
    elif d < s + 57:
        return 4
    elif d < s + 77:
        return 5
    elif d < s + 92:
        return 6
    else:
        return 7


def print_race_state(total_distance, race_pos) -> None:
    """Outputs the current state/stop of the race.
    Each horse is placed according to the distance they have travelled. In
    case some horses travelled the same distance, their numbers are printed
    on the same name"""

    # we dont want to modify the `race_pos` list, since we need
    # it later. Therefore we generating an interator from the list
    race_pos_iter = iter(race_pos)

    # race_pos is stored by last to first horse in the race.
    # we get the next horse we need to print out
    next_pos = next(race_pos_iter)

    # start line
    basic_print("XXXXSTARTXXXX")

    # print all 28 lines/unit of the race course
    for line in range(28):

        # ensure we still have a horse to print and if so, check if the
        # next horse to print is not the current line
        # needs iteration, since multiple horses can share the same line
        while next_pos is not None and line == total_distance[next_pos]:
            basic_print(f"{next_pos} ", end="")
            next_pos = next(race_pos_iter, None)
        else:
            # if no horses are left to print for this line, print a new line
            basic_print("")

    # finish line
    basic_print("XXXXFINISHXXXX")


def simulate_race(odds):
    num_horses = len(HORSE_NAMES)

    # in spirit of the original implementation, using two arrays to
    # track the total distance travelled, and create an index from
    # race position -> horse index
    total_distance = [0] * num_horses

    # race_pos maps from the position in the race, to the index of the horse
    # it will later be sorted from last to first horse, based on the
    # distance travelled by each horse.
    # e.g. race_pos[0] => last horse
    #      race_pos[-1] => winning horse
    race_pos = list(range(num_horses))

    basic_print("\n1 2 3 4 5 6 7 8")

    while True:

        # advance each horse by a random amount
        for i in range(num_horses):
            total_distance[i] += get_distance(odds[i])

        # bubble sort race_pos based on total distance travelled
        # in the original implementation, race_pos is reset for each
        # simulation step, so we keep this behaviour here
        race_pos = list(range(num_horses))
        for line in range(num_horses):
            for i in range(num_horses - 1 - line):
                if total_distance[race_pos[i]] < total_distance[race_pos[i + 1]]:
                    continue
                race_pos[i], race_pos[i + 1] = race_pos[i + 1], race_pos[i]

        # print current state of the race
        print_race_state(total_distance, race_pos)

        # goal line is defined as 28 units from start
        # check if the winning horse is already over the finish line
        if total_distance[race_pos[-1]] >= 28:
            return race_pos

        # this was not in the original BASIC implementation, but it makes the
        # race visualization a nice animation (if the terminal size is set to 31 rows)
        time.sleep(1)


def print_race_results(race_positions, odds, bets, player_names) -> None:
    """Print the race results, as well as the winnings of each player"""

    # print the race positions first
    basic_print("THE RACE RESULTS ARE:")
    for position, horse_idx in enumerate(reversed(race_positions), start=1):
        line = f"{position} PLACE HORSE NO. {horse_idx} AT {odds[horse_idx]}:1"
        basic_print("")
        basic_print(line)

    # followed by the amount the players won
    winning_horse_idx = race_positions[-1]
    for idx, name in enumerate(player_names):
        (horse, amount) = bets[idx]
        if horse == winning_horse_idx:
            basic_print("")
            basic_print(f"{name} WINS ${amount * odds[winning_horse_idx]}")


def main_loop(player_names, horse_odds) -> None:
    """Main game loop"""

    while True:
        print_horse_odds(horse_odds)
        bets = get_bets(player_names)
        final_race_positions = simulate_race(horse_odds)
        print_race_results(final_race_positions, horse_odds, bets, player_names)

        basic_print("DO YOU WANT TO BET ON THE NEXT RACE ?")
        one_more = basic_input("YES OR NO")
        if one_more.upper() != "YES":
            break


def main() -> None:
    # introduction, player names and horse odds are only generated once
    introduction()
    player_names = setup_players()
    horse_odds = setup_horses()

    # main loop of the game, the player can play multiple races, with the
    # same odds
    main_loop(player_names, horse_odds)


if __name__ == "__main__":
    main()

File: 50_Horserace/java/Horserace.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Scanner;

/**
 * HORSERACE
 * <p>
 * Converted from BASIC to Java by Aldrin Misquitta (@aldrinm)
 */
public class Horserace {

    private static final String[] horseNames = {
            "JOE MAW",
            "L.B.J.",
            "MR.WASHBURN",
            "MISS KAREN",
            "JOLLY",
            "HORSE",
            "JELLY DO NOT",
            "MIDNIGHT"
    };
    public static final int MAX_DISTANCE = 28;
    public static final int NUM_HORSES = 8;

    public static void main(String[] args) {
        printHeader();

        Scanner scanner = new Scanner(System.in);
        Random random = new Random();

        printHelp(scanner);

        List<String> betNames = readBetNames(scanner);

        boolean donePlaying = false;
        while (!donePlaying) {

            int[] odds = generateOdds(random);
            int sumOdds = Arrays.stream(odds).sum();
            printOdds(sumOdds, odds);

            Map<String, Bet> bets = takeBets(scanner, betNames);

            var horsePositions = runRace(horseNames.length, sumOdds, odds, random);

            printRaceResults(horsePositions, bets, sumOdds, odds);

            donePlaying = readDonePlaying(scanner);
        }
    }

    private static int[] generateOdds(Random random) {
        int[] odds = new int[NUM_HORSES];
        for (int i = 0; i < NUM_HORSES; i++) {
            odds[i] = (int) (10 * random.nextFloat() + 1);
        }
        return odds;
    }

    private static void printOdds(int R, int[] D) {
        System.out.printf("%n%-28s%-14s%-14s%n%n", "HORSE", "NUMBER", "ODDS");
        for (int n = 0; n < horseNames.length; n++) {
            System.out.printf("%-28s% -14d%.6f :1%n", horseNames[n], n + 1, ((float) R / D[n]));
        }
    }

    private static boolean readDonePlaying(Scanner scan) {
        System.out.println("DO YOU WANT TO BET ON THE NEXT RACE ?");
        System.out.print("YES OR NO? ");
        String choice = scan.nextLine();
        return !choice.equalsIgnoreCase("YES");
    }

    /**
     * Simulate the race run, returning the final positions of the horses.
     */
    private static int[] runRace(int numberOfHorses, int sumOdds, int[] odds, Random random) {
        int[] positionChange = new int[numberOfHorses];

        System.out.println();
        System.out.println("1 2 3 4 5 6 7 8");

        int totalDistance = 0;
        int[] currentPositions = new int[NUM_HORSES];
        int[] horsePositions = new int[NUM_HORSES];

        while (totalDistance < MAX_DISTANCE) {
            System.out.println("XXXXSTARTXXXX");

            for (int i = 0; i < numberOfHorses; i++) {
                horsePositions[i] = i + 1;
                positionChange[i] = calculatePositionChanges(sumOdds, odds[i], random);
                currentPositions[i] += positionChange[i];
            }

            sortHorsePositionsBasedOnCurrent(currentPositions, horsePositions);

            totalDistance = currentPositions[horsePositions[7] - 1];

            boolean raceFinished = false;
            int i = 0;
            while (i < NUM_HORSES && !raceFinished) {
                int distanceToNextHorse = currentPositions[(horsePositions[i] - 1)] - (i < 1 ? 0 : currentPositions[(horsePositions[i - 1] - 1)]);
                if (distanceToNextHorse != 0) {
                    int a = 0;
                    while (a < distanceToNextHorse && !raceFinished) {
                        System.out.println();
                        if (currentPositions[horsePositions[i] - 1] >= MAX_DISTANCE) {
                            raceFinished = true;
                        }
                        a++;
                    }
                }

                if (!raceFinished) {
                    System.out.print(" " + horsePositions[i] + " "); // Print horse number
                }
                i++;
            }

            if (!raceFinished) {
                //Print additional empty lines
                for (int a = 0; a < MAX_DISTANCE - totalDistance; a++) {
                    System.out.println();
                }
            }

            System.out.println("XXXXFINISHXXXX");
            System.out.println("\n");
            System.out.println("---------------------------------------------");
            System.out.println("\n");
        }

        return horsePositions;
    }

    /**
     * Sorts the horsePositions array in place, based on the currentPositions of the horses.
     * (bubble sort)
     */
    private static void sortHorsePositionsBasedOnCurrent(int[] currentPositions, int[] horsePositions) {
        for (int l = 0; l < NUM_HORSES; l++) {
            int i = 0;
            /*
            uses a do-while instead of a for loop here, because in BASIC
            a FOR I=1 TO 0 causes at least one execution of the loop
            */
            do {
                if (currentPositions[horsePositions[i] - 1] >= currentPositions[horsePositions[i + 1] - 1]) {
                    int h = horsePositions[i];
                    horsePositions[i] = horsePositions[i + 1];
                    horsePositions[i + 1] = h;
                }
                i++;
            } while (i < (7 - l));
        }
    }

    private static int calculatePositionChanges(int r, int d, Random random) {
        int positionChange = (int) (100 * random.nextFloat() + 1);

        if (positionChange < 10) {
            positionChange = 1;
        } else {
            int s = (int) ((float) r / d + 0.5);
            if (positionChange < (s + 17)) {
                positionChange = 2;
            } else if (positionChange < s + 37) {
                positionChange = 3;
            } else if (positionChange < s + 57) {
                positionChange = 4;
            } else if (positionChange < s + 77) {
                positionChange = 5;
            } else if (positionChange < s + 92) {
                positionChange = 6;
            } else {
                positionChange = 7;
            }
        }

        return positionChange;
    }

    private static void printRaceResults(int[] m, Map<String, Bet> bets, int r, int[] d) {
        System.out.println("THE RACE RESULTS ARE:");
        int z9 = 1;
        for (int i = 7; i >= 0; i--) {
            int f = m[i];
            System.out.println();
            System.out.println(z9 + " PLACE HORSE NO. " + f + " AT " + (r / d[f - 1]) + ":1");
            z9++;
        }
        bets.forEach((betName, bet) -> {
            if (bet.horseNumber == m[7]) {
                int n = bet.horseNumber;
                System.out.println();
                System.out.printf("%s WINS $ %.2f %n", bet.betName, ((float) r / d[n]) * bet.amount);
            }
        });
    }

    private static Map<String, Bet> takeBets(Scanner scanner, List<String> betNames) {
        Map<String, Bet> bets = new HashMap<>();
        System.out.println("--------------------------------------------------");
        System.out.println("PLACE YOUR BETS...HORSE # THEN AMOUNT");
        for (String betName : betNames) {
            boolean validInput = false;
            while (!validInput) {
                int horseNumber = readInt(betName, scanner);//Q in the original
                double betAmount = readDouble("?", scanner); //P in the original
                if (betAmount < 1 || betAmount > 100000) {
                    System.out.println("  YOU CAN'T DO THAT!");
                } else {
                    bets.put(betName, new Bet(betName, horseNumber, betAmount));
                    validInput = true;
                }
            }
        }

        return bets;
    }

    private static void printHelp(Scanner scanner) {
        System.out.print("DO YOU WANT DIRECTIONS");

        String directions = readChoice(scanner);

        if (!directions.equalsIgnoreCase("NO")) {
            System.out.println("UP TO 10 MAY PLAY.  A TABLE OF ODDS WILL BE PRINTED.  YOU");
            System.out.println("MAY BET ANY + AMOUNT UNDER 100000 ON ONE HORSE.");
            System.out.println("DURING THE RACE, A HORSE WILL BE SHOWN BY ITS");
            System.out.println("NUMBER.  THE HORSES RACE DOWN THE PAPER!");
            System.out.println();
        }
    }

    private static String readChoice(Scanner scanner) {
        System.out.print("? ");
        return scanner.nextLine();
    }

    private static int readInt(String prompt, Scanner scanner) {
        System.out.print(prompt);
        while (true) {
            System.out.print("? ");
            String input = scanner.nextLine();
            try {
                return Integer.parseInt(input);
            } catch (NumberFormatException e) {
                System.out.println("!NUMBER EXPECTED - RETRY INPUT LINE");
            }
        }
    }

    private static double readDouble(String prompt, Scanner scanner) {
        System.out.print(prompt);
        while (true) {
            System.out.print("? ");
            String input = scanner.nextLine();
            try {
                return Double.parseDouble(input);
            } catch (NumberFormatException e) {
                System.out.println("!NUMBER EXPECTED - RETRY INPUT LINE");
            }
        }
    }

    private static List<String> readBetNames(Scanner scanner) {
        int c = readInt("HOW MANY WANT TO BET ", scanner);
        System.out.println("WHEN ? APPEARS,TYPE NAME");
        List<String> names = new ArrayList<>();
        for (int i = 1; i <= c; i++) {
            System.out.print("? ");
            names.add(scanner.nextLine());
        }

        return names;
    }

    private static void printHeader() {
        System.out.println("                                               HORSERACE");
        System.out.println("                            CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
        System.out.println("\n\n");
        System.out.println("WELCOME TO SOUTH PORTLAND HIGH RACETRACK");
        System.out.println("                      ...OWNED BY LAURIE CHEVALIER");
    }

    private static class Bet {
        String betName;
        int horseNumber;
        double amount;

        public Bet(String betName, int horseNumber, double amount) {
            this.betName = betName;
            this.horseNumber = horseNumber;
            this.amount = amount;
        }

        @Override
        public String toString() {
            return "Bet{" +
                    "betName='" + betName + '\'' +
                    ", horseNumber=" + horseNumber +
                    ", amount=" + amount +
                    '}';
        }
    }
}

